### ISSUE TITLE
Dynamic dataloader with accelerator .prepare()

### ISSUE THREAD

**Issue by sleepymalc** [huggingface/accelerate#3706]
I need to design training algorithms that will dynamically add data to the dataset. I wonder what the best practice is for integrating such a dynamic dataloader with accelerator's `.prepare()` function?

It seems like some "uncommon" training dataloader iteration will lead to unexpected behavior.
> E.g., manually iterating dataloader with `iter` prepared by accelerator accelerator's step will be disturbed, and gradient accumulation will behave abnormally.

Hence, I'm worried that these uncommon dataloader operation during the training will disturb the accelerator's internal state.
